plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,615 @@
+{
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = {
+		PXI_HID_DEV_OTA_INPUT_REPORT_ID,
+		0x0,
+	};
+	if (!fu_udev_device_pread_full (FU_UDEV_DEVICE (self),
+					port, res, (FU_PXI_DEVICE_NOTTFY_RET_LEN + 1) - port,
+					error))
+		return FALSE;
+	if (status != NULL) {
+		if (!fu_common_read_uint8_safe (res, sizeof(res), 0x1,
+						status, error))
+			return FALSE;
+	}
+	if (checksum != NULL) {
+		if (!fu_common_read_uint16_safe (res, sizeof(res), 0x3,
+						 checksum, G_LITTLE_ENDIAN, error))
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_object_create (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	guint8 opcode = 0;
+	/* request */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+	fu_byte_array_append_uint32 (req, chk->address, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint32 (req, chk->data_sz, G_LITTLE_ENDIAN);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* check object create success or not */
+	if (!fu_pxi_device_wait_notify (self, 0x0, &opcode, NULL, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwWrite opcode invalid %02x",
+			     opcode);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_write_payload (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	g_byte_array_append (req, chk->data, chk->data_sz);
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					   req->data, req->len, error);
+}
+
+static gboolean
+fu_pxi_device_write_chunk (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	guint32 prn = 0;
+	guint16 checksum = fu_pxi_device_calculate_checksum (chk->data, chk->data_sz);
+	guint16 checksum_tmp = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* send create fw object command */
+	if (!fu_pxi_device_fw_object_create (self, chk, error))
+		return FALSE;
+
+	/* write payload */
+	chunks = fu_chunk_array_new (chk->data, chk->data_sz, chk->address,
+				     0x0, self->mtu_size);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk2 = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_payload (self, chk2, error))
+			return FALSE;
+		prn++;
+		if (prn >= self->prn_threshold) {
+			guint8 opcode = 0;
+			if (!fu_pxi_device_wait_notify (self, 0x0, &opcode, &checksum_tmp, error))
+				return FALSE;
+			if (opcode != FU_PXI_DEVICE_CMD_FW_WRITE) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_READ,
+					     "FwWrite opcode invalid %02x",
+					     opcode);
+				return FALSE;
+			}
+			prn = 0;
+		}
+	}
+	self->checksum  +=  checksum;
```

Comment:
I don't think you want to use self->checksum; this is supposed to be the checksum as reported from FwInitNew and it's a bit weird to see it used here.

Reply:
Our checksum is accumulation ,so I need it

Reply:
our checksum is accumulation, so we need sum with last checksum. The last checksum is from FwInitNew. It could be zero or other(resume transmit)

Reply:
I've merged your changes into origin/wip/pixart-rf -- can you test that please. You need to submit future pull requests against that branch as it's quite different to your wip/pixart-rf branch now.